File: src/bindings/scala/api/src/main/scala/com/ctc/omega_edit/api/Session.scala
/*
 * Copyright 2021 Concurrent Technologies Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ctc.omega_edit.api

import com.ctc.omega_edit.api.Change.Result
import com.ctc.omega_edit.api.Session.OverwriteStrategy

import java.nio.file.Path
import scala.util.Try

/**
  * The top level type in OmegaEdit, maintains the data and all instances related to it.
  * Provides mutators and viewport factory methods.
  */
trait Session {
  def size: Long
  def isEmpty: Boolean = size == 0

  def numChanges: Long
  def numCheckpoints: Long
  def numUndos: Long
  def numViewports: Long
  
  def callback: Option[SessionCallback]

  def eventInterest: Int
  def eventInterest_=(eventInterest: Int): Unit

  def insert(s: String, offset: Long): Result
  def insert(b: Array[Byte], offset: Long): Result

  def overwrite(s: String, offset: Long): Result
  def overwrite(b: Array[Byte], offset: Long): Result

  def delete(offset: Long, len: Long): Result
  def view(offset: Long, size: Long): Viewport

  def undoLast(): Result
  def redoUndo(): Result

  def clearChanges(): Result

  def getLastChange(): Result

  def getLastUndo(): Result

  def viewCb(offset: Long, size: Long, cb: ViewportCallback, eventInterest: Int): Viewport
  def findChange(id: Long): Option[Change]

  def save(to: Path): Try[Path]
  def save(to: Path, overwrite: OverwriteStrategy): Try[Path]

  def search(pattern: String, offset: Long, length: Option[Long] = None, caseInsensitive: Boolean = false, limit: Option[Long] = None): List[Long]

  def getSegment(offset: Long, length: Long): Option[Segment]

  def pauseSessionChanges(): Result
  def resumeSessionChanges(): Result
}

object Session {
  sealed trait OverwriteStrategy
  object OverwriteStrategy {
    case object OverwriteExisting extends OverwriteStrategy
    case object GenerateFilename extends OverwriteStrategy
  }
}

File: src/rpc/client/ts/src/session.ts
/*
 * Copyright (c) 2021 Concurrent Technologies Corporation.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ByteFrequencyProfileRequest,
  CreateSessionRequest,
  ObjectId,
  SaveSessionRequest,
  SearchRequest,
  SegmentRequest,
} from './omega_edit_pb'
import { Empty } from 'google-protobuf/google/protobuf/empty_pb'
import { getClient } from './settings'
const client = getClient()

export function createSession(
  path: string | undefined,
  session_id_desired: string | undefined
): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    let request = new CreateSessionRequest()
    if (session_id_desired !== undefined && session_id_desired.length > 0)
      request.setSessionIdDesired(session_id_desired)
    if (path !== undefined && path.length > 0) request.setFilePath(path)
    client.createSession(request, (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('createSession error: ' + err.message)
      }
      return resolve(r.getSessionId())
    })
  })
}

export function destroySession(id: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    client.destroySession(new ObjectId().setId(id), (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('destroySession error: ' + err.message)
      }
      return resolve(r.getId())
    })
  })
}

export function saveSession(
  session_id: string,
  file_path: string,
  overwrite: boolean
): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    client.saveSession(
      new SaveSessionRequest()
        .setSessionId(session_id)
        .setFilePath(file_path)
        .setAllowOverwrite(overwrite),
      (err, r) => {
        if (err) {
          console.log(err.message)
          return reject('saveSession error: ' + err.message)
        }
        return resolve(r.getFilePath())
      }
    )
  })
}

export function getComputedFileSize(session_id: string): Promise<number> {
  return new Promise<number>((resolve, reject) => {
    client.getComputedFileSize(new ObjectId().setId(session_id), (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('getComputedFileSize error: ' + err.message)
      }
      return resolve(r.getComputedFileSize())
    })
  })
}

export function pauseSessionChanges(session_id: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    client.pauseSessionChanges(new ObjectId().setId(session_id), (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('pauseSessionChanges error: ' + err.message)
      }
      return resolve(r.getId())
    })
  })
}

export function resumeSessionChanges(session_id: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    client.resumeSessionChanges(new ObjectId().setId(session_id), (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('resumeSessionChanges error: ' + err.message)
      }
      return resolve(r.getId())
    })
  })
}

export function unsubscribeSession(session_id: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    client.unsubscribeToSessionEvents(
      new ObjectId().setId(session_id),
      (err, r) => {
        if (err) {
          console.log(err.message)
          return reject('unsubscribeSession error: ' + err.message)
        }
        return resolve(r.getId())
      }
    )
  })
}

export function getSegment(
  session_id: string,
  offset: number,
  length: number
): Promise<Uint8Array> {
  return new Promise<Uint8Array>((resolve, reject) => {
    client.getSegment(
      new SegmentRequest()
        .setSessionId(session_id)
        .setOffset(offset)
        .setLength(length),
      (err, r) => {
        if (err) {
          console.log(err.message)
          return reject('getSegment error: ' + err.message)
        }
        return resolve(r.getData_asU8())
      }
    )
  })
}

export function getSessionCount(): Promise<number> {
  return new Promise<number>((resolve, reject) => {
    client.getSessionCount(new Empty(), (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('getSessionCount error: ' + err.message)
      }
      return resolve(r.getCount())
    })
  })
}

export function profileSession(
  session_id: string,
  offset: number | undefined,
  length: number | undefined
): Promise<number[]> {
  return new Promise<number[]>((resolve, reject) => {
    let request = new ByteFrequencyProfileRequest().setSessionId(session_id)
    if (offset !== undefined && offset >= 0) request.setOffset(offset)
    if (length !== undefined && length > 0) request.setLength(length)
    client.getByteFrequencyProfile(request, (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('searchSession error: ' + err.message)
      }
      return resolve(r.getFrequencyList())
    })
  })
}

// Given a computed profile, return the total number of bytes in the 7-bit ASCII range
export function numAscii(profile: number[]): number {
  return profile.slice(0, 128).reduce((accumulator, current) => {
    return accumulator + current
  }, 0)
}

export function searchSession(
  session_id: string,
  pattern: string | Uint8Array,
  is_case_insensitive: boolean | undefined,
  offset: number | undefined,
  length: number | undefined,
  limit: number | undefined
): Promise<number[]> {
  return new Promise<number[]>((resolve, reject) => {
    let request = new SearchRequest()
      .setSessionId(session_id)
      .setPattern(typeof pattern == 'string' ? Buffer.from(pattern) : pattern)
      .setIsCaseInsensitive(is_case_insensitive ?? false)
    if (offset !== undefined && offset >= 0) request.setOffset(offset)
    if (length !== undefined && length > 0) request.setLength(length)
    if (limit !== undefined && limit > 0) request.setLimit(limit)
    client.searchSession(request, (err, r) => {
      if (err) {
        console.log(err.message)
        return reject('searchSession error: ' + err.message)
      }
      return resolve(r.getMatchOffsetList())
    })
  })
}
